### ISSUE TITLE
[Feature]: Serverside filtering and sort for Tables

### ISSUE THREAD

**Issue by lublak** [appsmithorg/appsmith#10486]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Binding of the data from the Filter and Sort fields to the data sources.

### Why should this be worked on?

Currently, the filtering and sorting is client-side.
Especially with large amounts of data, I use serverside pageination.
But now it would be practical to integrate the data in sql as with serverside pagination. Table.sortCol and Table.filterJson next to Table.search and Table.pageNo would be really nice.

**Comment by Nikhil-Nandagopal** [comment]
@lublak sort is already available. You can bind {{Table1.sortOrder.column}}. Server-side filtering is still missing today

**Comment by somangshu** [comment]
@lublak this might be possible today even on the server side using the onSort event, which can be used to call an api, We also have an internal property {{Table1.sortOrder}} which gives you the column name and the order in which the user is asking you to sort, this is set to null when no column is set on sort. 

<img width="1159" alt="Screenshot 2022-01-24 at 4 21 03 PM" src="https://user-images.githubusercontent.com/11089579/150770023-5419ee6e-3466-4dc4-bba6-e0e60c2292e4.png">

@Nikhil-Nandagopal do you think this solves the problem? This is not well documented though. cc @Pranay105 @dilippitchika

**Comment by Nikhil-Nandagopal** [comment]
@somangshu it does but the same needs to be possible for filters as well

**Comment by somangshu** [comment]
@Nikhil-Nandagopal agreed, we have an [open issue](https://github.com/appsmithorg/appsmith/issues/5398) for that, not yet prioritised, we might we able to get it up once we have the v2 for table.

**Comment by somangshu** [comment]
@Tooluloope I remember you working on the earlier feature set, Any comments on this, What is the effort required? Did we push this back for a reason?

**Comment by lublak** [comment]
Yes for sorting there is it available (but are we missing the serverside switch like the serverside pagination to disable all client side searching?)
Filtering is still completely missing.

**Comment by dilippitchika** [comment]
Hey @lublak could you expand on the switch part? You have events called onSearchTextChanged and onSort. 

If you bind a query to it, it will be triggered with the events making it server side, if you don't it will be done on the client side.  Is this part of the table events unclear?

Agreed on the filtering part.

**Comment by lublak** [comment]
@dilippitchika
![firefox_AmE1SbxE8Q](https://user-images.githubusercontent.com/44057030/151438981-be442f7b-4876-4076-87b0-54a9bb448fd0.png)
 the Sort switch is also missing

**Comment by dilippitchika** [comment]
Hey @lublak the reason we have "Server side Pagination" as a switch is because it needs extra properties for it to be configured and used.

For server side sort, we don't have a switch because once you configure the **onSort action** to bind to a query it is server side sort. You don't need an extra switch for it. When there is no query bound to onSort it uses client side sort.

The switch is only needed for server side if there are some extra properties which are needed to be filled.

**Comment by lublak** [comment]
@dilippitchika and what about "Enable client side search"? Why is this needed?

**Comment by Nikhil-Nandagopal** [comment]
@lublak that is used to disable the client-side search when server-side search is applied way so that the client-side does not override the server-side search

**Comment by Tooluloope** [comment]
@lublak Here's the issue to understand why we added the property
https://github.com/appsmithorg/appsmith/issues/5444

**Comment by dilippitchika** [comment]
Today, we already expose all the filter properties using Table1.filters, we need to add a new onFilterUpdate action to support this.

**Comment by lublak** [comment]
@Nikhil-Nandagopal @Tooluloope thanks for the answer :) i understood now.
It's just a bit confusing when it's there for searching and not for sorting. I'm sorry I didn't understand it right away.

**Comment by somangshu** [comment]
@dilippitchika in case there is enough clerity here with what needs to be implemented, Please move this to the backlog and unassign yourself

**Comment by dilippitchika** [comment]
Action items for this feature
1. Add .filters into the autocomplete for users to know it can be accessed.
2. Add a new onFilterUpdate which detects the changes made in the filters and triggers this action.

@somangshu now that we have 2 versions of tables how to uptake this change?

**Comment by Tooluloope** [comment]
Should this be moved to Table WIdget V2 or remain in Enhancement?

**Comment by dilippitchika** [comment]
Will check and get back tomorrow, @Tooluloope

**Comment by dilippitchika** [comment]
@Tooluloope we should do this in V1 as well, checked with @sbalaji1192. Please discuss with him before picking it up.

**Comment by somangshu** [comment]
@dilippitchika considering the previous discussion, can you drop the summary here and add what is required to resolve the actionable, We will resolve this internally.

> Action desired behavior for this feature
> - Add .filters into the autocomplete for users to know it can be accessed.
> - Add a new onFilterUpdate which detects the changes made in the filters and triggers this action.

**Comment by dilippitchika** [comment]
@somangshu @Tooluloope please review and move this to product backlog if clear

**Comment by Nikhil-Nandagopal** [comment]
@dilippitchika instead of a new property, can we just rename "Enable client side search" to "Enable client-side filtering" and use it for both search and filter because I don't see much of a use case to have search on and filter off.

**Comment by dilippitchika** [comment]
## Stats
| Stat | Values |
| ------------- | ------------- |
| Reach     | 200 |
| Effort (months)  | 0.5 |

**Comment by somangshu** [comment]
Will be handled after #9135 release

**Comment by acron0** [comment]
This is a deal breaker for us, unfortunately. Any ideas when it might come around?

**Comment by somangshu** [comment]
Hey @acron0, The dependency on this have been cleared, but we still havent prioritized this. 

While I talk about this internally, can you explain your use case, Maybe we can suggest some workaround.

cc @dilippitchika

**Comment by dilippitchika** [comment]
[Internal] PRD -https://www.notion.so/appsmith/Supporting-server-side-filtering-445d8f060bf14c5b98e62dcbcd835162

**Comment by andrewdietekoki** [comment]
1. TL;DR
The design task was to accommodate for:
- Switching On Server-Side Filtering property
- Allowing a Column to be filterable or not.
- Filterable Property reflected in Filter Dialog.

2. [Screens Design](https://www.figma.com/file/1L37f1ahbinjfBWg4tDS82/Server-Side-Filter-for-the-Table-Widget?node-id=2145%3A111270&t=KPixZynjvahYG3jk-1)

3. [Prototype](https://www.figma.com/proto/1L37f1ahbinjfBWg4tDS82/Server-Side-Filter-for-the-Table-Widget?page-id=2146%3A168063&node-id=2146%3A220811&viewport=676%2C366%2C0.03&scaling=min-zoom&starting-point-node-id=2146%3A220811)

**Comment by m1ker1n** [comment]
Hi, is there any progress about the issue? I would like to use it too

**Comment by somangshu** [comment]
Hey @m1ker1n, today we expose a field called as `filters` in the table widget that allows you to get the filters that are applied and maybe also use that in queries if required. 

<img width="1295" alt="Screenshot 2023-07-12 at 3 58 37 PM" src="https://github.com/appsmithorg/appsmith/assets/11089579/91676e14-09fa-4322-b42b-6bc909160abe">

**Comment by m1ker1n** [comment]
> Hey @m1ker1n, today we expose a field called as `filters` in the table widget that allows you to get the filters that are applied and maybe also use that in queries if required.
> 
> <img alt="Screenshot 2023-07-12 at 3 58 37 PM" width="1295" src="https://user-images.githubusercontent.com/11089579/252939858-91676e14-09fa-4322-b42b-6bc909160abe.png">

How should I observe filters change to run new query with new parameters?

**Comment by somangshu** [comment]
@m1ker1n apologies for the confusion, discussed this internally and found out that we need an action like onFilterUpdate on the table widget to complete this loop. This will not work as expected otherwise.

**Comment by eyyupguner** [comment]
@somangshu any upcoming news about the issue in near future

**Comment by somangshu** [comment]
@eyyupguner can you please help me understand what you are looking for. 
We haven’t yet prioritized this issue internally.

**Comment by eyyupguner** [comment]
@somangshu i have implemented the same logic with input boxes for my all columns but its hard to implement if you have lots of tables in your project and hard to manage their names etc.

**Comment by m1ker1n** [comment]
We created button "apply filters" which takes these filters and handles them. So user must click one more time after he chose them in popup to receive filtrated data from server

**Comment by LagunaElectric** [comment]
@rahulbarwal @carinanfonseca Do we have an ETA for this?

**Comment by yonatang-cyera** [comment]
Hey @rahulbarwal  @carinanfonseca  , we'd very much like to utilize your built-in filters and connect them to our APIs for backend functionality - developing our own filters with "select" widgets creates a very high complexity and a UI which is not clean. It adds an additional level of maintenance and a complex UX. Also - this is for the most simple "AND" use case with exact match - for "contains" and "OR" conditions we'll need an even more complicated UI. 
Your built-in filters are very user friendly and we'd like to utilize them with server side pagination. Using them today with server-side is unusable since they only filter the current page. 
Same goes to other capabilities such as "download", but filters is our current focus.
It'll be very appreciated if you can provide an ETA.
An alternative can be an ability to "fork" your table as a custom component and a full control over its code.

Thanks!

![image](https://github.com/user-attachments/assets/3f82e047-f628-46e6-a45a-5156cb935cc7)

**Comment by carinanfonseca** [comment]
Hello @yonatang-cyera! I would love to have a short call with you about this topic to better understand how critical this topic is for you and to discuss some key points. Is this possible?

If yes, feel free to book a time that works best for you [here](https://calendly.com/carina-neves-fonseca/appsmith-feedback).
I am looking forward to speaking with you!
